File: typescript/common/http.ts

import logger from "./logger";
import { ResponseError } from "./errors";

export const StatusCodes = {
    OK: 200,
    CREATED: 201,
} as const;
  

async function validateResponse(response: Response, expectedStatusCodes: number[], logResponses: boolean = false) {
    if (expectedStatusCodes.includes(response.status)) {
        if (logResponses) {
            logger.debug(`Received expected response (${response.status}) from ${response.url}`);
        }
    } else {
        const responseText = await response.text();
        throw new ResponseError(`Received unexpected response (${response.status}) from ${response.url}\n${responseText}`);
    }
}

export function HttpRequest(expectedStatusCodes: number[], logResponses: boolean = false) {
    return function (target: any, propertyKey: string, descriptor: PropertyDescriptor) { 
        const originalMethod = descriptor.value;

        descriptor.value = async function(...args: any[]): Promise<number> {
            let response: Response = await originalMethod.apply(this, args);
            await validateResponse(response, expectedStatusCodes, logResponses);
            return await response.json();
        }

        return descriptor;
    };
}

File: python/core/http.py
import httpx
import functools
from typing import List, Callable, Awaitable
from python.core.exception import ResponseError
from python.core.logger import logger

try:
    from typing import ParamSpec
except ImportError:
    from typing_extensions import ParamSpec

P = ParamSpec("P")


class StatusCodes:
    OK = 200
    CREATED = 201


def validate_response(response: httpx.Response, expected_status_codes: List[int], log_responses: bool = False):
    request_endpoint = f"{response.request.method} {response.request.url}"
    if response.status_code in expected_status_codes:
        if log_responses:
            logger.debug(f"Received expected response ({response.status_code}) from {request_endpoint}")
    else:
        msg = f"Received unexpected response ({response.status_code}) from {request_endpoint}\n{response.text}"
        raise ResponseError(msg)


def http_request(expected_status_codes: List[int], log_responses: bool = False):
    def decorator(func: Callable[[P], Awaitable[httpx.Response]]) -> Callable[[P], Awaitable[dict]]:
        @functools.wraps(func)
        async def wrapper(*args: P.args, **kwargs: P.kwargs) -> dict:
            response: httpx.Response = await func(*args, **kwargs)
            validate_response(response, expected_status_codes=expected_status_codes, log_responses=log_responses)
            return response.json()

        return wrapper

    return decorator
